_examples/real-world-examples/server-sent-events/server/event_handlers.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/router/middleware"
+	"github.com/ThreeDotsLabs/watermill/pubsub/gochannel"
+)
+
+const (
+	PostCreatedTopic = "post-created"
+	PostUpdatedTopic = "post-updated"
+	FeedUpdatedTopic = "feed-updated"
+)
+
+func SetupMessageRouter(
+	feedsStorage FeedsStorage,
+	logger watermill.LoggerAdapter,
+) (message.Publisher, message.Subscriber, error) {
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
+	if err != nil {
+		return nil, nil, err
+	}
+	router.AddMiddleware(middleware.Recoverer)
+
+	pubsub := gochannel.NewGoChannel(gochannel.Config{}, logger)
+
+	publishEvents := func(ctx context.Context, tags []string) (messages []*message.Message, err error) {
+		defer func() {
+			if err == nil {
+				logger.Info("Updated posts in feeds", nil)
+			} else {
+				logger.Error("Error in handler", err, nil)
+			}
+		}()
+
+		for _, tag := range tags {
+			logger.Info("Producing event", watermill.LogFields{"tag": tag})
+			event := FeedUpdated{
+				Name:       tag,
+				OccurredAt: time.Now().UTC(),
+			}
+
+			payload, err := json.Marshal(event)
+			if err != nil {
+				return nil, err
+			}
+
+			newMessage := message.NewMessage(watermill.NewUUID(), payload)
+
+			messages = append(messages, newMessage)
+		}
+
+		return messages, nil
+	}
+
+	router.AddHandler(
+		"on-post-created",
+		PostCreatedTopic,
+		pubsub,
+		FeedUpdatedTopic,
+		pubsub,
+		func(msg *message.Message) (messages []*message.Message, err error) {
+			defer func() {
+				if err == nil {
+					logger.Info("Event handler on-post-created executed successfully", nil)
+				} else {
+					logger.Error("Error in on-post-created event handler", err, nil)
+				}
+			}()
+
+			event := PostCreated{}
+			err = json.Unmarshal(msg.Payload, &event)
+			if err != nil {
+				return nil, err
+			}
+
+			logger.Info("Adding post", watermill.LogFields{"post": event.Post})
+
+			if len(event.Post.Tags) > 0 {
+				for _, tag := range event.Post.Tags {
+					logger.Info("Adding tag", watermill.LogFields{"tag": tag})
+					err = feedsStorage.Add(msg.Context(), tag)
+					if err != nil {
+						return nil, err
+					}
+
+				}
+
+				err = feedsStorage.AppendPost(msg.Context(), event.Post)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			return publishEvents(msg.Context(), event.Post.Tags)
```

Comment:
the name publishEvents is a bit missleading, it suggests that it is publishing something, and it doesn't say what events they are. maybe generate feed updated events? or better, generate read model updated events (because feed was updated earlier, read model is updated now)

Reply:
c4eba92ae51e6cb3656a235ee014acfb11293212